File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.py
from collections import deque
def bfs(sj, si, ej, ei):
    q=deque()
    v=[0]*n

    q.append((sj,si))

    while q:
        cj,ci = q.popleft()
        if abs(cj-ej)+abs(ci-ei)<=1000: # 목적지 도달가능
            return 'happy'
        
        # 미방문 모든 편의점좌표: 범위내인지 체크
        for i in range(n):
            if v[i]==0:     # 미방문 편의점
                nj,ni = lst[i]
                if abs(cj-nj)+abs(ci-ni)<=1000: # 범위내
                    q.append((nj,ni))
                    v[i]=1
    return 'sad'

t=int(input())
for testCase in range(t):
    n=int(input())
    sj, si = map(int, input().split())
    lst = []

    for _ in range(n):
        tj, ti = map(int, input().split())
        lst.append((tj,ti))
    ej, ei = map(int, input().split())

    answer=bfs(sj,si,ej,ei)
    print(answer)

File: 백준/Gold/9205. 맥주 마시면서 걸어가기/맥주 마시면서 걸어가기.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine());
		for (int tc = 1; tc <= t; tc++) {
			int n = Integer.parseInt(br.readLine());
			int[][] points = new int[n + 2][2];
			
			for (int i = 0; i < n + 2; i++) {
				StringTokenizer st = new StringTokenizer(br.readLine());
				points[i][0] = Integer.parseInt(st.nextToken());
				points[i][1] = Integer.parseInt(st.nextToken());
			}
			
			boolean[][] flag = new boolean[n + 2][n + 2];
			for (int i = 0; i < n + 2; i++) {
				for (int j = 0; j < n + 2; j++) {
					if (getDistance(points[i], points[j]) <= 1000) flag[i][j] = true;
				}
			}

			for (int k = 0; k < n + 2; k++) {
				for (int i = 0; i < n + 2; i++) {
					for (int j = 0; j < n + 2; j++) {
						if (flag[i][k] && flag[k][j]) flag[i][j] = true;
					}
				}				
			}
			
			String answer;
			if (flag[0][n + 1]) answer = "happy";
			else answer = "sad";
			System.out.println(answer);
		}
	}
	
	static int getDistance(int[] p1, int[] p2) {
		return Math.abs(p1[0] - p2[0]) + Math.abs(p1[1] - p2[1]);
	}
}
